test/e2e/operator_test.go

```diff
@@ -4241,6 +4243,108 @@ func waitForIngressControllerCondition(t *testing.T, cl client.Client, timeout t
 	return err
 }
 
+func assertEIPAllocationDeleted(t *testing.T, svc *ec2.EC2) {
+	t.Helper()
+	cleanupEIPAllocations(t, svc)
+}
+
+func cleanupEIPAllocations(t *testing.T, svc *ec2.EC2) {
+	t.Log("Release unassociated EIPs ")
+
+	clusterName, err := getClusterName(infraConfig)
+	if err != nil {
+		t.Fatalf("failed to get cluster name due to error: %v", err)
+	}
+
+	vpcID, err := getVPCID(clusterName, svc)
+	if err != nil {
+		t.Fatalf("failed to get VPC ID due to error %s", err)
+	}
+
+	tagKeyEIP := fmt.Sprintf("NE1674-%s", os.Getenv("USER"))
+	tagValueEIP := clusterName
+
+	// Describe addresses to get unassociated EIPs
+	describeAddressesInput := &ec2.DescribeAddressesInput{}
+	describeAddressesOutput, err := svc.DescribeAddresses(describeAddressesInput)
+	if err != nil {
+		t.Fatalf("failed to describe addresses, %v", err)
+	}
+
+	var unassociatedEIPs []struct {
+		PublicIp     string
+		AllocationId string
+	}
+
+	for _, address := range describeAddressesOutput.Addresses {
+		if address.AssociationId == nil {
+			// Check if the EIP has the specified tag key and value
+			hasTag := false
+			for _, tag := range address.Tags {
+				if *tag.Key == tagKeyEIP && *tag.Value == tagValueEIP {
+					hasTag = true
+					break
+				}
+			}
+			if hasTag {
+				// Check if the EIP is associated with a network interface in the specific VPC
+				if address.NetworkInterfaceId != nil {
+					describeNetworkInterfacesInput := &ec2.DescribeNetworkInterfacesInput{
+						NetworkInterfaceIds: []*string{address.NetworkInterfaceId},
+					}
+					describeNetworkInterfacesOutput, err := svc.DescribeNetworkInterfaces(describeNetworkInterfacesInput)
+					if err != nil {
+						t.Fatalf("failed to describe network interfaces, %v", err)
+					}
+
+					for _, networkInterface := range describeNetworkInterfacesOutput.NetworkInterfaces {
+						if networkInterface.VpcId != nil && *networkInterface.VpcId == vpcID {
+							unassociatedEIPs = append(unassociatedEIPs, struct {
+								PublicIp     string
+								AllocationId string
+							}{
+								PublicIp:     aws.StringValue(address.PublicIp),
+								AllocationId: aws.StringValue(address.AllocationId),
+							})
+						}
+					}
+				} else {
+					// If there's no associated network interface, assume it's an unassociated EIP
+					unassociatedEIPs = append(unassociatedEIPs, struct {
+						PublicIp     string
+						AllocationId string
+					}{
+						PublicIp:     aws.StringValue(address.PublicIp),
+						AllocationId: aws.StringValue(address.AllocationId),
+					})
+				}
```

Comment:
I'm a bit confused with this logic. I know you can't release an associated EIP, I've seen that error before, but this logic skips EIPs that have our tag BUT are associated. That could create a EIP leak no?

If we created it, we *must* delete it, or it's this cleanup should fail the test, even if it's still associated. I can understand doing a preliminary poll waiting for all of our Tagged EIPs to become unassociated, then throwing a `t.Error` if one of our EIP allocation never unassociates (but still continuing to release what EIPs we can).

Regardless, this cleanup logic should fail if we find a EIP with our tag, and it never becomes unassociated. Also, this logic block would be more readable as a helper function such as `isEIPAssociated` or `eipAssociated`.

Reply:
Until the Ingress Controller is deleted the EIP wont be released. If the Ingress Controller fails to be deleted we will get an error. Also we can't release the EIP which are associated with a Ingress Controller.

Reply:
anyway that's a good point thanks! I understood what needs to be done.

Reply:
done

Reply:
1. check if service lb was deleted
2. poll until the eip with tag and value are unassociated.
3. Release unassociated eips.

Reply:
So, with the current logic I am checking if the service was deleted, then poll the unassociated eips and release them.